src/commands/utils.rs

```diff
@@ -62,3 +71,93 @@ where
     let s = Vec::from_hex(&s).map_err(de::Error::custom)?;
     consensus::deserialize(&s).map_err(de::Error::custom)
 }
+
+/// Select coins for spend.
+///
+/// Returns the selected coins and the change amount, which could be zero.
+///
+/// `candidate_coins` are the coins to consider for selection.
+///
+/// `base_tx` is the transaction to select coins for. It should be without any inputs
+/// and without a change output, but with all non-change outputs added.
+///
+/// `change_txo` is the change output to add if needed (with any value).
+///
+/// `feerate_vb` is the minimum feerate (in sats/vb).
+///
+/// `min_fee` is the minimum fee (in sats) that the selection must have.
+///
+/// `max_sat_weight` is the maximum size difference (in vb) of
+/// an input in the transaction before and after satisfaction.
+pub fn select_coins_for_spend(
+    candidate_coins: &[CandidateCoin],
+    base_tx: bitcoin::Transaction,
+    change_txo: bitcoin::TxOut,
+    feerate_vb: u64,
+    min_fee: u64,
+    max_sat_weight: usize,
+) -> Result<(Vec<Coin>, bitcoin::Amount), InsufficientFunds> {
+    let max_input_weight = TXIN_BASE_WEIGHT + max_sat_weight as u32;
+    let candidates: Vec<Candidate> = candidate_coins
+        .iter()
+        .map(|cand| Candidate {
+            input_count: 1,
+            value: cand.coin.amount.to_sat(),
+            weight: max_input_weight,
+            is_segwit: true, // We only support receiving on Segwit scripts.
+        })
+        .collect();
+    // Transaction base weight is calculated from transaction with no inputs and no change output.
+    let base_weight = base_tx.weight().to_wu() as u32;
+    let long_term_feerate = FeeRate::from_sat_per_vb(LONG_TERM_FEERATE_VB as f32);
+    let target = Target {
+        value: base_tx.output.iter().map(|o| o.value).sum(),
+        feerate: FeeRate::from_sat_per_vb(feerate_vb as f32),
+        min_fee,
+    };
+    let drain_weights = DrainWeights {
+        output_weight: {
+            let mut tx_with_change = base_tx;
+            tx_with_change.output.push(change_txo);
+            tx_with_change.weight().to_wu() as u32 - base_weight
+        },
+        spend_weight: max_input_weight,
+    };
+    // Change policy ensures any change output is not too small and that transaction waste is reduced.
+    let change_policy =
+        change_policy::min_value_and_waste(drain_weights, DUST_OUTPUT_SATS, long_term_feerate);
+
+    let mut selector = CoinSelector::new(&candidates, base_weight);
+    // Select any mandatory candidates.
+    for (i, cand) in candidate_coins.iter().enumerate() {
+        if cand.must_select {
+            selector.select(i);
+        }
+    }
```

Comment:
I think it's OK as the `select()` method uses the index based on the original candidates slice rather than how they are currently ordered, but moving this code up and adding a comment will make this clearer.